contracts/infrastructure/dapp/ParaswapFilter.sol

```diff
@@ -33,55 +35,208 @@ interface IParaswap {
+        (IParaswap.MegaSwapSellData memory sell) = abi.decode(_data[4:], (IParaswap.MegaSwapSellData));
+        return hasValidBeneficiary(_wallet, sell.beneficiary) &&
+            hasTradableToken(sell.path[0].path[sell.path[0].path.length - 1].to) && 
+            hasValidMegaPath(sell.path);
+    }
+
+    function hasAuthorisedCallees(
+        address _augustus,
+        address[] memory _callees,
+        uint256[] memory _startIndexes,
+        bytes calldata _data
+    )
+        internal
+        view
+        returns (bool)
+    {
+        // _data = {sig:4}{six params:192}{exchangeDataOffset:32}{...}
+        // we add 4+32=36 to the offset to skip the method sig and the size of the exchangeData array
+        uint256 exchangeDataOffset = 36 + abi.decode(_data[196:228], (uint256)); 
+        for(uint256 i = 0; i < _callees.length; i++) {
+            bytes calldata slicedExchangeData = _data[exchangeDataOffset+_startIndexes[i] : exchangeDataOffset+_startIndexes[i+1]];
+            address spender = Utils.recoverSpender(_callees[i], slicedExchangeData);
+            if(!authoriser.isAuthorised(_augustus, spender, _callees[i], slicedExchangeData)) {
+                return false;
+            }
+        }
+        return true;
     }
 
-    function hasValidBeneficiary(address _wallet, bytes calldata _data) internal pure returns (bool) {
-        (address beneficiary) = abi.decode(_data[228:], (address)); // skipping 4 + 7*32 = 228 bytes
-        return (beneficiary == address(0) || beneficiary == _wallet);
+    function hasValidBeneficiary(address _wallet, address _beneficiary) internal pure returns (bool) {
+        return (_beneficiary == address(0) || _beneficiary == _wallet);
     }
 
-    function hasTradableToken(bytes calldata _data) internal view returns (bool) {
-        (, address _destToken) = abi.decode(_data[4:], (address, address));   
+    function hasTradableToken(address _destToken) internal view returns (bool) {
         if(_destToken == ETH_TOKEN) {
             return true;
         }
         (bool success, bytes memory res) = tokenPriceRegistry.staticcall(abi.encodeWithSignature("isTokenTradable(address)", _destToken));
         return success && abi.decode(res, (bool));
     }
 
-    function hasValidExchangeAdapters(bytes calldata _data) internal view returns (bool) {
-        // Note: using uint256[5] instead of (address, address, uint, uint, uint) to avoid "stack too deep" issues
-        (,IParaswap.Path[] memory path) = abi.decode(_data[4:], (uint256[5], IParaswap.Path[]));
+    function hasValidExchangeAdapters(IParaswap.Path[] memory _path) internal view returns (bool) {
         (bool success,) = dexRegistry.staticcall(
-            abi.encodeWithSignature("verifyExchangeAdapters((address,uint256,(address,address,uint256,bytes,uint256)[])[])", path)
+            abi.encodeWithSignature("verifyExchangeAdapters((address,uint256,(address,address,uint256,bytes,uint256)[])[])", _path)
```

Comment:
Feels a bit unfortunate to have both a DappRegistry and a DexRegistry... would be nice to being able to get rid of the latter.

Reply:
We could get rid of `DexRegistry` by using filters that authorise only a specific account (in this case AugustusSwapper) to use them. To avoid having wasting gas on `IFilter(filter).isValid()` calls, we could add the rule that if a filter address is the address of the account for which the authorisation is requested, the authorisation should be granted:
```
function isAuthorised(address _wallet, address _spender, address _to, bytes calldata _data) external view override returns (bool) {
        /// ....
                    if(filter == address(0) || filter == _wallet || IFilter(filter).isValid(_wallet, _spender, _to, _data)) {
                        return true;
                    }
        }
```
That would let us verify the exchange adapters at roughly the same cost as we pay now.

Reply:
Actually for adapters I think it's fine to hardcode their list in the filter